File: Stroids/src/Pulse.cpp
/*
 * Pulse.cpp    Feb 27, 2014
 * AUTHOR: Caleb Reister
 * DEV ENV: Eclipse 4 CDT Linux AMD64
 * DESCRIPTION: Defines the "bullets" fired
 *  by the ship in the asteroids game implementation.
 */

//#define NDEBUG
#include <cassert>
#include "Pulse.hpp"

/* Pulse()
 * Initializes pulse data for testing purposes.
 */
Pulse::Pulse()
{
    //cfg = &stroidConfig;
    timeToLive = 250;
    radius = 0;
    location = ZERO;
    setVelocity(0, 0, 0);
    setAngle(0);
}

/* Pulse(Vect2d loc, float ang)
 * Constructor that passes location and angle data, used to fire a pulse from the ship.
 * Sets a constant velocity.
 *
 * Vect2d loc - the starting location of the pulse
 * float ang - the starting angle of the pulse
 */
Pulse::Pulse(Vect2d loc, float ang)
{
    Vect2d vel = util::deg2slope(ang);
    timeToLive = cfg::getInst()->read["LASER"]["PULSE_LIFE"].as_int();
    radius = 1;
    location = loc;
    setAngle(ang);

    vel *= 100;
    setVelocity(vel, cfg::getInst()->read["SHIP"]["SPEED"].as_float() + 2);
    pulse.setRotation(getAngle());
    pulse.setPosition(loc.x, loc.y);
    pulse.setSize(sf::Vector2f(30, 3));
}

/* void checkScreenEdge()
 * Checks to see if the pulse has reached the edge of the screen.
 * Prevents it from going to the other side.
 */

/* void hit()
 * Sets timeToLive to 0, invoking the draw function will
 * cause the pulse to disappear.
 */
void Pulse::hit()
{
    timeToLive = 0;
}

/* void isDead()
 * Checks to see if the pulse has hit anything.
 */
bool Pulse::isDead()
{
    assert(timeToLive >= 0);
    if (timeToLive == 0)
        return true;
    else
        return false;
}

/* draw(sf::RenderWindow& win)
 * Draws the pulse on the screen where and when appropriate.
 *
 * sf::RenderWindow& win - the window in which to draw the pulse.
 */
void Pulse::draw(sf::RenderWindow& win)
{
    Vect2d loc = getLocation();

    pulse.setPosition(loc.x, loc.y);
    pulse.setFillColor(sf::Color(255, 255, 255));

    win.draw(pulse);
    timeToLive--;
    updateLocation();
}

File: StroidsPy/src/Pulse.py
##@file Pulse.py
#@brief Defines the "bullets" fired by the ship in the
# asteroids game implementation.

import sfml as sf
import SpaceObject
import util

##@class Pulse
#@brief This class is a SpaceObject that simulates laser pulses.
class Pulse(SpaceObject.SpaceObject):
    radius = 1
    _pulse = sf.RectangleShape()
    _timeToLive = 250
    _color = sf.Color(255, 255, 255)
    #################################################
    def __init__(self, loc, ang):
        self.setAngle(ang)
        self._loc = loc
        
        vel = util.deg2slope(ang)
        self.setVelocity(vel, SHIP_SPEED + 2)
        
        self._pulse.set_position(self._loc[0], self._loc[1])
        self.pulse.set_rotation(self.getAngle())
        self.pulse.set_size(sf.Vector2(30, 3))
        
        #color setting code goes here
    ###################################################
    def hit(self):
        self._timeToLive = 0
    ####################################################
    def isDead(self):
        assert(self._timeToLive >= 0)
        if self._timeToLive == 0:
            return True
        else:
            return False
    ###################################################
    def draw(self, win):
        self._pulse.set_position(self._loc[0], self._loc[1])
        self._pulse.set_fill_color(sf.Color(self._color))
        
        win.draw(self._pulse)
        self._timeToLive -= 1
        self.updateLocation()
        
        if not EDGE_DEATH:
            self._boundFix()
    